level06/src/FirstApp/AppM.hs

```diff
@@ -1,83 +1,73 @@
--- foo :: ReaderT r m a
--- When run, becomes:
--- foo :: r -> m a
---
 -- First, let's clean up our (Conf,FirstAppDB) with an application Env type. We
--- will add a general purpose logging function as well
+-- will add a general purpose logging function as well. Remember that functions
+-- are values, we're able to pass them around and place them on records like any
+-- other type.
 data Env = Env
   { envLoggingFn :: Text -> AppM ()
   , envConfig    :: Conf
   , envDB        :: FirstAppDB
   }
 
--- Lets crack on and define a newtype wrapper for our ReaderT, this will save us
--- having to write out the full ReaderT definition for every function that uses
--- it.
-
--- Our ReaderT will only contain the Env, and our base monad will be IO, leave
--- the return type polymorphic so that it will work regardless of what is
--- being returned from the functions that will use it. Using a newtype (in
--- addition to the useful type system) means that it is harder to use a
--- different ReaderT when we meant to use our own, or vice versa. In such a
--- situation it is extremely unlikely the application would compile at all,
--- but the name differences alone make the confusion less likely.
-
--- Because we're using a newtype, all of the instance definitions for ReaderT
--- would normally not apply. However, because we've done nothing but create a
--- convenience wrapper for our ReaderT, it is not difficult for GHC to
--- automatically derive instances on our behalf.
-
--- With the 'GeneralizedNewtypeDeriving' pragma at the top of the file, we
--- will be able to derive these instances automatically.
-newtype AppM a = AppM
-  { unAppM :: ReaderT Env IO a }
-  deriving ( Functor
-           , Applicative
-           , Monad
-           , MonadReader Env
-           , MonadIO
-           )
-
--- Below is a helper function that will take the requirements for our ReaderT,
--- an Env, and the (AppM a) that is the action to be run with the given Env.
-
--- Our AppM must first be 'unwrapped', the newtype definition we wrote gives us
--- that function:
--- > unAppM :: AppM a -> ReaderT Env IO a
---
--- Then we run the ReaderT, using:
--- > runReaderT :: ReaderT r m a -> r -> m a
--- ~
--- > runReaderT :: ReaderT Env IO a -> Env -> IO a
---
--- Composing them (runReaderT . unAppM) we are left with:
--- > Env -> IO a
+-- It would be nice to remove the need to pass around our Env to every function
+-- that needs it. Wouldn't it be great to have our functions run where we could
+-- simply ask for the current Env?
 --
--- We have an Env so that leaves us with the:
--- > IO a
--- and we're done.
+-- We can create this by wrapping a function in a newtype like so:
+newtype AppM a = AppM ( Env -> IO a )
+-- This means that any function where we want to use the Env will now return
+-- this data structure. Which is a function that needs an Env. We can then write
+-- a `Monad` instance for this type that we use to pass the Env from function to
+-- function.
```

Comment:
I think this comment could be clearer, something like:

> Our application will be built around functions of the shape `Env -> IO a`, or, functions which (may) need an `Env` as input, and will return results of type `a` using `IO`. Not all functions using `AppM` will need the Env, and not all will need to use IO, but if we're consistent, we can abstract the pattern of "functions which might need an Env and might need to perform IO " so that we don't have to care which functions do and don't need these _effects_; all operations will be passed the Env and they can ignore it if they don't need it, and all are capable of IO, but don't necessarily need to perform it.

Reply:
The broad message is good but I don't like parts of this, because in other languages any function might perform IO but doesn't necessarily need to perform it. We run screaming from those languages for a reason.

May be worth foreshadowing that we can ask for `MonadReader Env m` or `MonadIO m` instead of demanding all the effects which we may not use.